subsys/bluetooth/host/iso.c

```diff
@@ -1096,3 +1105,427 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < big->num_bis; i++) {
+		if (!big->bis[i]) {
+			BT_DBG("BIG BIS[%d] not initialized", i);
+			return -EINVAL;
+		}
+	}
+
+	/* They all have the same QOS dir so we can just check the first */
+	broadcaster = big->bis[0]->qos->dir == BT_ISO_CHAN_QOS_OUT;
+
+	if (broadcaster) {
+		err = hci_le_terminate_big(big);
+	} else {
+		err = hci_le_big_sync_term(big);
+	}
+
+	if (err) {
+		BT_DBG("Could not terminate BIG %d", err);
+	} else {
+		bigs[big->handle] = NULL;
+		cleanup_big(big);
+	}
+
+	return err;
+}
+
+void hci_le_big_sync_established(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_big_sync_established *evt = (void *)buf->data;
+	uint16_t big_handle = sys_le16_to_cpu(evt->big_handle);
+	struct bt_iso_big *big = bigs[big_handle];
+
+	if (evt->status || evt->num_bis != big->num_bis) {
+		cleanup_big(big);
+		for (int i = 0; i < big->num_bis; i++) {
+			struct bt_iso_chan *bis = big->bis[i];
+
+			if (bis->ops->disconnected) {
+				bis->ops->disconnected(bis);
+			}
+		}
+		return;
+	}
+
+	for (int i = 0; i < big->num_bis; i++) {
+		struct bt_iso_chan *bis = big->bis[i];
+		uint16_t bis_handle = sys_le16_to_cpu(evt->handle[i]);
+
+		bis->conn->handle = bis_handle;
+
+		bt_iso_chan_set_state(bis, BT_ISO_CONNECTED);
+
+		if (bis->ops->connected) {
+			bis->ops->connected(bis);
+		}
+	}
+
+	/* TODO: Deal with the rest of the fields in the event,
+	 * if it makes sense
+	 */
+}
+
+void hci_le_big_sync_lost(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_big_sync_lost *evt = (void *)buf->data;
+	uint16_t handle = sys_le16_to_cpu(evt->big_handle);
+	struct bt_iso_big *big = bigs[handle];
+
+	BT_DBG("BIG[%u] sync lost", big->handle);
+
+	bigs[big->handle] = NULL;
+	big_disconnect(big);
+	cleanup_big(big);
+}
+
+static int hci_le_big_create_sync(const struct bt_le_per_adv_sync *sync,
+				  struct bt_iso_big *big,
+				  const struct bt_iso_big_sync_param_t *param)
+{
+	struct bt_hci_cp_le_big_create_sync *req;
+	struct net_buf *buf;
+	int err;
+	uint8_t bit_idx = 0;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_BIG_CREATE_SYNC,
+				sizeof(*req) + big->num_bis);
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	req = net_buf_add(buf, sizeof(*req) + big->num_bis);
+	req->big_handle = sys_cpu_to_le16(big->handle);
+	req->sync_handle = sys_cpu_to_le16(sync->handle);
+	req->encryption = param->encryption;
+	if (req->encryption) {
+		memcpy(req->bcode, param->bcode, sizeof(req->bcode));
+	} else {
+		memset(req->bcode, 0, sizeof(req->bcode));
+	}
+	req->mse = param->mse;
+	req->sync_timeout = sys_cpu_to_le16(param->sync_timeout);
+	req->num_bis = big->num_bis;
+	/* Transform from bitfield to array */
+	for (int i = 0; i < 0x1F; i++) {
+		if (param->bis_bitfield & BIT(i)) {
+			if (bit_idx == big->num_bis) {
+				BT_DBG("BIG cannot contain %u BISes",
+				       bit_idx + 1);
+				return -EINVAL;
+			}
+			req->bis[bit_idx++] = i;
+		}
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_BIG_CREATE_SYNC, buf, NULL);
+	if (err) {
```

Comment:
Superfluous lines - err is returned in any case

Reply:
Good catch